README.md
# BrainDamaged

BrainDamaged is a project that builds the multiple Projects.

* BrainDamagedVM a 32-bit Bytecode Virtual Machine
* BrainDamagedAssembler compiles Assembly into Bytecode
* ~~BrainFuckCompiler a BrainFuck to BrainDamagedVM Bytecode compiler.~~

# BrainDamaged VM

Location: `bdvm/`  
Version: 0.5.0

## Architecture

* All Instructions are 32-Bit.
* Maximum Stack size is 4096 variables (4096 * 32 bits ~= 16KB)
* Maximum Program size is dynamic with no upper limits, the default allocation is 4096 (4096 * 32 bits ~= 16KB)

+ This VM has a Memory size from at least 32 KB.
  * ~16 KB is addressable Stack memory
  * ~16 KB is at least program code memory

Observing the Memory on Linux, we use about 1721 KB with a 3 instruction program (0x01, RSI, HLT).

## Building

We expect this to build and run on these OS Configurations, this however is not guaranteed:

| OS | Version | Arch | Toolkit |
|-----|--------|------|---------|
| Linux | \>= 2.6.11 | amd64 | LLVM 9 |
| Windows | \>= 10 (1909) | amd64 | MSVC 14 |

Steps:
1. Get Source Code
2. Get Submodules
3. Build

#### Get Source Code

```
$ git clone https://gitlab.atvg-studios.com/atvg-studios/braindamaged
$ cd braindamaged
```

#### Get Submodules

```
$ git submodule init
```

#### Build

```
$ mkdir build && cd build
$ cmake .. && make
```

## Instruction Set

* pc = Program Counter (Current executing Instruction)
* sp = Stack Pointer (Current top of Stack)
* tos = Top of Stack

| Name | Bytecode | ARGC | Description | Stack Action | Since |
|------|-----------|----------------|-------------|--------------|-------|
| HLT | 0x40000000 | - | Halts the VM | - | 0.1.0 |
| NOP | 0x40000001 | - | Does Nothing | - | 0.1.0 |
|  -  | 0x40000002 | - | RESERVED | - | - |
|  -  | 0x40000003 | - | RESERVED | - | - |
|  -  | 0x40000004 | - | RESERVED | - | - |
| ADD | 0x40000005 | - | Adds two numbers | POPs twice (sp-2), puts result on tos (sp++) | 0.1.0 |
| SUB | 0x40000006 | - | Subtracts two numbers | POPs twice (sp-2), puts result on tos (sp++) | 0.1.0 |
| MUL | 0x40000007 | - | Multiplies two numbers | POPs twice (sp-2), puts result on tos (sp++) | 0.1.0 |
| DIV | 0x40000008 | - | Divides two numbers | POPs twice (sp-2), puts result on tos (sp++) | 0.1.0 |
| DUP | 0x40000009 | - | Duplicates top of stack | Copies top of stack into tos (sp++) | 0.1.0 |
| JE  | 0x4000000A | 1 | Skips `x` instructions when tos is 1 | POPs once (sp--) | 0.1.0/0.3.0 |
| JNE | 0x4000000B | 1 | Skips `x` instructions when tos is 0 | POPs once (sp--) | 0.1.0/0.3.0 |
| CMP | 0x4000000C | - | Compares two values and puts 0 or 1 | POPs twice (sp-2), puts result on tos (sp++) | 0.1.0 |
| POP | 0x4000000D | - | Removes from tos | POPs once (sp--) | 0.1.0 |
| SWP | 0x4000000E | - | Switches tos | POPs twice (sp-2), puts both switched to stack (sp+2) | 0.1.0 |
| INC | 0x4000000F | - | Increments tos | - | 0.2.0 |
| DEC | 0x40000010 | - | Decrements tos | - | 0.2.0 |
| JMP | 0x40000011 | 1 | Skips `x` instructions | - | 0.2.0/0.3.0 |
| SPI | 0x40000012 | - | Increments sp | (sp++) | 0.3.0 |
| SPD | 0x40000013 | - | Decrements sp | (sp--) | 0.3.0 |
| SPC | 0x40000014 | - | Resets sp | (sp = first non-null location from end of stack) | 0.3.0 |
| LDI | 0x40000015 | 1 | Adds a value to the stack | Puts value on tos (sp++) | 0.5.0 |
| - | - | - | - | - | - |
| PSI | 0x40000051 | - | Prints tos as Int to stdout | POPs tos (sp--) | 0.1.0 |
| PSC | 0x40000052 | - | Prints tos as ASCII Char to stdout | POPs tos (sp--) | 0.1.0 |
| RSI | 0x40000053 | - | Reads from stdin to tos as Int | Puts to tos (sp++) | 0.1.0 |
| RSC | 0x40000054 | - | Reads from stdin to tos as ASCII Char | Puts to tos (sp++) | 0.1.0 |
| RSW | 0x40000055 | - | Reads from stdin to stack as ASCII until end of line (Max 1024 bytes) | Puts to tos (sp += in bytes / 3) | 0.3.0 |
| PSW | 0x40000056 | - | Prints to stdout from stack in reverse as ASCII | POPs from tos (sp -= (in bytes / 3) + 2) | 0.3.0 |

# BrainDamaged Assembler

Location: `bdasm/`  
Version: 0.3.0

# Disclaimer
**Please don't expect this to work. This is the first Bytecode Compiler and Bytecode VM I wrote in C++.**

# Attributes

BrainDamaged logo made by [Freepik](https://www.flaticon.com/authors/freepik) from [Flaticon](https://www.flaticon.com/)

# License

Just don't bother yet. Do with the code what you want, if you event want to touch it ..

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

* [BrainDamagedVM](#bdvm)
* [BrainDamagedAssembler](#bdasm)
* [BrainFuckCompiler](#bdc)

# <a name="bdvm"></a>BrainDamagedVM

## [Unreleased]

### Changed
- 16-Bit instead of 32-bit instructions
- Jumps are fixed and not static
- Memory can hold 8, 16, 32, 64 bit integers and 64 bit floats (floats are yet to be supported in instructions)

## [0.5.0] - 2020-05-04
### Changed
- Many Internal changes
- Correctly SP out of bounds errors
- Loading integers can now only be done using the LDI Instruction

## [0.4.0] - 2020-04-17
### Added
- Support for Windows (TODO: Make the Program name output dynamic just like on Linux)
- Safety checks for each Instruction that depends on a SP > 0 < PC_BEGIN

### Removed
- Instruction: DBG - This was moved to VM flag -d
- Instruction: DDM - This is always enabled when the VM gets started with the -d flag


## [0.3.0] - 2020-04-16
### Added
- Instruction: DDM - Debug Dump Memory displays a hex memory dump when DBG is enabled
- Instruction: RSW - Read Stack Wide reads from stdin until a end of line was given
- Instruction: PSW - Print Stack Wide prints a string if sp is txt_end (0x03) and a txt_begin is found (0x02) (txt_begin and txt_end are from ASCII)
- Instruction: SPI - Stack Pointer Increment, increments the SP by 1
- Instruction: SPD - Stack Pointer Decrement, decrements the SP by 1
- Instruction: SPC - Stack Pointer Clear, Sets SP to the first spot from the back of stack memory (4096) which is not null

### Changed
- Program counter starts at 4096 making space for 4096 stack variables (about 15 kb)
- Jump instructions (JMP, JNE, JE) now take a 32-bit int as argument

## [0.2.0] - 2020-04-16
### Added
- Instruction: INC - Increments value at stack pointer
- Instruction: DEC - Decrements value at stack pointer
- Instruction: JMP - Adds stack pointer to program counter to skip x instructions (relative jumps)

## [0.1.0] - 2020-04-16
### Added
- First Implementation of BrainDamagedVM

# <a name="bdasm"></a>BrainDamagedAssembler

## [Unreleased]

### Changed
- Store instructions as 16-Bit
- Apply name change of basic math instructions

## [0.3.0] - 2020-05-04
### Changed
- Loading integers can now only be done using the LDI Instruction

## [0.2.0] - 2020-04-17
### Added
- Support for Windows (TODO: Make the Program name output dynamic just like on Linux)

## [0.1.0] - 2020-04-16
### Added
- First Implementation of BrainDamagedAssembler

# <a name="bdc"></a>BrainFuckCompiler

## [Unreleased]
## [0.1.0] - 2020-04-17
### Added
- Tried to do a first implementation (It is more complicated then though xD)
- We now have the basic commands except for `[]` 
- Optional VM commands with `-x` (Currently no Extensions due to removal of DBG and DDM Bytecode)
- Support for Windows (TODO: Make the Program name output dynamic just like on Linux)

bdvm/include/Bytecode.hpp
//
// Created by atjontv on 4/16/20.
//

#ifndef BRAINDAMAGED_BYTECODE_HPP
#define BRAINDAMAGED_BYTECODE_HPP

#define HLT 0x40000000
#define NOP 0x40000001
#define ADD 0x40000005
#define SUB 0x40000006
#define MUL 0x40000007
#define DIV 0x40000008
#define DUP 0x40000009
#define JE  0x4000000A // Jumps to SP if SP is 0
#define JNE 0x4000000B // Jumps to SP if SP is 1
#define CMP 0x4000000C // Compares SP-1 with SP
#define POP 0x4000000D
#define SWP 0x4000000E
#define INC 0x4000000F
#define DEC 0x40000010
#define JMP 0x40000011 // Jumps to SP
#define SPI 0x40000012
#define SPD 0x40000013
#define SPC 0x40000014 // Sets SP to 0
#define LDI 0x40000015
#define PSI 0x40000051
#define PSC 0x40000052
#define RSI 0x40000053
#define RSC 0x40000054
#define RSW 0x40000055
#define PSW 0x40000056

#endif //BRAINDAMAGED_BYTECODE_HPP
